Fix recurring payments on Indian CCs

ismaeldcom opened the pull request:
Fixes #2935

### Changes proposed in this Pull Request
With the new regulations to process recurring payments for cards issued by Indian banks, we need to generate and save a mandate ID, which will be attached to every renewal order.

- Generate mandate params from subscriptions details. Implement `get_mandate_params_for_order` and call it within `process_payment` and `process_payment_for_order` using `$additional_api_parameters`.
- Save the returned mandate ID from payment intent succeed webhook on the order meta as `_stripe_mandate_id`.
- Attach mandate to renewal orders if required. Implement `get_mandate_param_for_renewal_order` and call it within `scheduled_subscription_payment` using `$additional_api_parameters`.
- Add an order note about the required authorization of the payment by the merchant if required. Implement `maybe_add_customer_notification_note` and call it within `process_payment_for_order` with the new `$processing` property.
- Add `$processing` property to `WC_Payments_API_Intention` to be able to access `customer_notification`.
- Update intent failed webhook handler to support intents without charges, using `last_payment_error` instead. Add messages for the new codes, and include the Stripe error message with our default one.

### Known issues
- Changing payment method on a subscription which have a mandate will result in a failed renovation. Using another Indian CC will result in an error by mandate not matching the payment method. And using any other card will also fail because we will attach the initial mandate.
  - **Possible fix**: Update the mandate on payment method change, removing it if it's empty.
- Mandates are created in a pending state and transition to active after 60-90 seconds. If we attempt to renew a subscription within that time frame, it will fail and add an order note with the following message: `Error: Only active mandates can be used with PaymentIntents.`
  - **Possible fix**: Delay the subscription renovation until the mandate is ready to be used, listening to `mandate.updated` webhook.

### Testing instructions

> **Note**: Stripe simulates the async nature of the renewal process delaying 5-10 minutes both success and failure responses.

#### Requirements
- Place orders in `INR` using Multi-currency or updating your store currency.
- Be sure you're listening to Stripe webhooks as the implementation relies on them. (`npm run listen`)
- Create a subscription product below `₹15,000 INR` and another one over it.

#### WC Subscriptions – Single subscription
- Enable WooCommerce Subscriptions and be sure you have early renewals enabled.
- Place and order with the following cards and subscription amounts.
  - 4000003560000123 – **less than** ₹15k
  - 4000003560000123 – **greater than** ₹15k
  - 4000003560000297 – any amount
  - 4000003560000248 – any amount
  - 4000003560000263 – any amount
- Wait a minute for the mandate to be active.
- Go to **WooCommerce → Status → Scheduled Actions → Pending** and filter by `woocommerce_scheduled_subscription_payment`.
- Run all the hooks where `subscription_id` matches the 5 subscriptions you have just created.
- Wait patiently for Stripe webhook 🕙.
- Order notes should match the following table.

| *123 <15k | *123 >15k | *297 | *248 | *263 |
| - | - | - | - | - |
|<img width="272" alt="success" src="https://user-images.githubusercontent.com/7670276/203037793-eb468ead-0ac7-4925-82dd-a06a01684d49.png">|<img width="270" alt="success-over-15k" src="https://user-images.githubusercontent.com/7670276/203037780-3c213e65-9875-4b3a-b6b9-b8337d6c4feb.png">|<img width="272" alt="fail-cancel-pause-297" src="https://user-images.githubusercontent.com/7670276/203037788-532b47ca-7bc6-4fa3-b50d-e7c762ce5f32.png">|<img width="270" alt="fail-notification-248" src="https://user-images.githubusercontent.com/7670276/203037795-2c153421-3354-4c63-95f0-11bc9b719876.png">|<img width="270" alt="fail-mandate-cancelled-263" src="https://user-images.githubusercontent.com/7670276/203037797-0f5a4ad5-9a1c-4dbe-b53e-e155664dc590.png">|

#### WC Subscriptions – Multiple subscriptions
- Enable WooCommerce Subscriptions and be sure you have early renewals enabled.
- Place an order using card `4000003560000123` with multiple subscriptions with:
  - Same interval.
  - Different interval.
- Wait a minute for the mandate to be active.
- Go to **WooCommerce → Status → Scheduled Actions → Pending** and filter by `woocommerce_scheduled_subscription_payment`.
- Run all the hooks where `subscription_id` matches the 2 subscriptions you have just created.
- Wait patiently for Stripe webhook 🕙.
- Renewals should succeed.

#### UPE vs classic checkout
The previous testing instructions should work for both UPE and classic checkout. Please test it using the missing one.

#### WCPay Subscriptions
Does not apply.

### Screenshots
#### Known issues
| Pending mandate | Wrong mandate | Not matching mandate |
| - | - | - |
|<img width="271" alt="pending-mandate" src="https://user-images.githubusercontent.com/7670276/203037798-e22c0c45-605e-46d9-9d36-3a7a0a7d92dd.png">|<img width="272" alt="wrong-mandate" src="https://user-images.githubusercontent.com/7670276/203037803-7eabe811-1e03-41ef-96ee-fef980180cff.png">|<img width="270" alt="not-matching-mandate" src="https://user-images.githubusercontent.com/7670276/203037807-eb9e4069-7697-4f10-b07b-5b6270b942a3.png">|

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**
- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : [Testing instructions](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-5.2.0#fix-recurring-payments-on-indian-ccs--pr3011)
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

ismaeldcom committed:
Initial add intent mandate implementaiton

Due to the new additional factor authentication for all recurring transactions from Indian credit cards, we need to include a new parameter on every subscription.

ismaeldcom commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -706,4 +706,34 @@ public function restore_order_pay_var() {
+			$subscriptions = wcs_get_subscriptions_for_order( $order->get_id() );
+			$subscription  = reset( $subscriptions );
+			if ( $subscription ) {
+				$result['setup_future_usage']                                = 'off_session';
```
Appreciate any input here. I think we should apply it to every order with subscriptions, and probably not here.

ismaeldcom commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -706,4 +706,34 @@ public function restore_order_pay_var() {
+			$subscriptions = wcs_get_subscriptions_for_order( $order->get_id() );
+			$subscription  = reset( $subscriptions );
+			if ( $subscription ) {
+				$result['setup_future_usage']                                = 'off_session';
```
Appreciate any input here. I think we should apply it to every order with subscriptions, and probably not here.

dmallory42 replied:
I agree it makes sense to apply it to every Subscription order. As per the Stripe docs:

> You can still accept off-session payments with a card set up for on-session payments, but the bank is more likely to reject the off-session payment and require authentication from the cardholder.

It seems to me that since subscriptions will always have renewal payments, `off_session` makes the most sense here. Maybe we can attach it in a different place, such as when we set up the payment intent? I'm not really sure where the best place for this is.

ismaeldcom replied:
Thanks! I'll need to think more about where to put it. For now, we only set this in `class-wc-payments-api-client.php` if the payment method will be saved. But it isn't enough for this use case. Cause `setup_future_usage=off_session` is always required with e-mandates, and if we use an already saved card will not work.

ismaeldcom committed:
Include mandate id if exists

ismaeldcom commented on includes/admin/class-wc-rest-payments-webhook-controller.php:
```diff
@@ -264,6 +264,12 @@ private function process_webhook_payment_intent_succeeded( $event_body ) {
 		}
 
+		// Save mandate id necessary for some subscriptions renewal.
+		$mandate_id = $event_body['data']['object']['charges']['data'][0]['payment_method_details']['card']['mandate'] ?? null;
```
This is quite deep! I opted for the null coalescing operator instead of checking for the existence of each key. Apart from that, not sure if there is a better place to put it. The mandate id is only included in the payment_intent.succeed so we need to rely on the event to get it.

dmallory42 replied:
Even though it's very deep, I can't really think of a better way to do this.

ismaeldcom committed:
Handle multiple subscriptions on a single order

ismaeldcom committed:
Add order note about payment authorisation by customer

ismaeldcom commented:
I'm wondering if we should handle the payment intent failed response from Stripe to add another order note and provide some more details on why it failed, to the merchant.

For now, Stripe sent us 2 cards that fail with a specific decline code (mentioned in testing instructions). We could handle those 2, but to do that we will need to also update the server to forward the `payment_intent.payment_failed` to the client. AFAIK it isn't being forwarded for now.

I'll appreciate any thought here. I see it as a bit out of the scope of this PR. Should we implement it here? Create another issue? Or is there a reason to not do this?

cc @Automattic/sigma

ismaeldcom committed:
Add a comma to the order note text

ismaeldcom committed:
Fix current tests

Add mocks, and required methods and properties to helpers

ismaeldcom committed:
Fix param should be optional

daquinons commented:
> Another question about order notes involving greater changes and a bit out of the scope in this PR.
> 
> If the mandate becomes invalid the merchant will see the following error. Looks like we just append the error message from Stripe on subscription payment fails. And not sure if we should improve at least this message to provide more useful information. Something like `Payment failed due to a wrong mandate, you will need to ask the customer for a new on session payment...`
> 
> What do you all think? Should we create another issue for this?
> 
> <img alt="Screenshot 2021-10-07 at 11 11 58" width="269" src="https://user-images.githubusercontent.com/7670276/136355550-3beec1b4-ca97-4ba2-a4a1-6888607c05c8.png">

Yes, definitely this deserves an issue. From the places I've seen we add notes into the order, they are always very user friendly to add context to the merchant, so good idea to create a new issue! 👍

ismaeldcom committed:
Add new tests

dmallory42 reviewed (approved):
Tested and all working perfectly after we finished debugging, nice work! 👍 I think this is ready to 🚢  once tests are done.

My thoughts on the comments you raised are as follows:

> I'm wondering if we should handle the payment intent failed response from Stripe to add another order note and provide some more details on why it failed, to the merchant.

> For now, Stripe sent us 2 cards that fail with a specific decline code (mentioned in testing instructions). We could handle those 2, but to do that we will need to also update the server to forward the payment_intent.payment_failed to the client. AFAIK it isn't being forwarded for now.

> I'll appreciate any thought here. I see it as a bit out of the scope of this PR. Should we implement it here? Create another issue? Or is there a reason to not do this?

I think it would be great for us to provide a generalised order note for example "A request to authorise this payment failed with the following error code: `transaction_not_approved`", but I think given we want to get this fix out, it would be fine to do this in a follow up issue.

I also agree with David about handling the other error message an a more friendly way but again this can be done in a followup issue IMO.

reykjalin commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -706,4 +719,70 @@ public function restore_order_pay_var() {
+	 * @param WC_Order $order The subscription order.
+	 * @return array Params to be included or empty array.
+	 */
+	public function maybe_add_mandate_to_order_payment( WC_Order $order ):  array {
```
I think you may need to add some special handling for orders where the initial payment doesn't have the same value as any of the renewal payments 🤔 

I think I handled this correctly in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981/commits/864cbb8a40dd4f76dc4c6feea40cad83f8e64f05, but I wasn't able to test it [due to API issues](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981#issuecomment-939246499).

Kudos to @ricardo for spotting this problem in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981#issuecomment-939208232

ismaeldcom replied:
Thanks @ricardo and @reykjalin for pointing it out! I missed it too.

Addressed in 2e53900

reykjalin reviewed (changes requested):
This looks good to me! I don't see any differences between this PR and my own PR in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981.

@ricardo found an issue in my PR that I think you have to address in this PR as well, but otherwise this is looking good 👌

ismaeldcom committed:
Update renew note about customer authorization

ismaeldcom committed:
Handle order with other products and signup fee

htdat reviewed (commented):
Nitpick: @ismaeldcom - can you help add changelog entries for readme.txt and changelog.txt?

reykjalin reviewed (approved):
LGTM 🙂

ismaeldcom committed:
Handle UPE new card subscription too

With UPE we use another process_payment method where we need to also call the add mandate function.

ismaeldcom commented:
Found an issue with UPE testing it again today. The mandate options were not added to subscriptions using UPE and a new card. Because we use another process, creating the intent without confirm, where I initially added the new code.

For it to work we need to add the `payment_method_options` param to `update_intention` on server 1241-gh-Automattic/woocommerce-payments-server.

**To ensure this is properly covered, I'll appreciate a pair of hands testing it. Will be enough to just place a subscription order following the main testing instructions, but using a new card, you just need to remove it if they are already used.**

Until #3118 got merged, would be better to add something like `sleep( 10 );` to `process_webhook_payment_intent_succeeded` if you experience some issues with the order not being found on `payment_intent.succeeded`.

dmallory42 reviewed (approved):
Tested the updated PR and all working as expected. Nice job! :shipit:

haszari commented:
Tagging @Automattic/helix for a review as requested. Note I've removed the `pr: ready to merge` label as it conflicts with the `on hold` status and to avoid confusion or accidental merge.

aprea commented:
@haszari I don't think this PR is ready for review yet given the merge conflicts. It looks like there is a plan to resolve these and potentially tweak some other code:

>The PR is now a bit outdated, we will need to solve any conflict, update the code and double-check that everything works as expected.

>Update PR code, test all scenarios, and ensure it’s compatible with WC Pay Subscriptions.

Probably best to wait until these are completed before we review & test this.

daquinons commented:
> @haszari I don't think this PR is ready for review yet given the merge conflicts. It looks like there is a plan to resolve these and potentially tweak some other code:
> 
> > The PR is now a bit outdated, we will need to solve any conflict, update the code and double-check that everything works as expected.
> 
> > Update PR code, test all scenarios, and ensure it’s compatible with WC Pay Subscriptions.
> 
> Probably best to wait until these are completed before we review & test this.

Yep, thank you @haszari, we appreciate the help! 

@ismaeldcom will work on putting the PR on shape during the week and he'll ping you when it's ready for review 👍

haszari commented:
> @ismaeldcom will work on putting the PR on shape during the week and he'll ping you when it's ready for review 👍

@ismaeldcom feel free to request a review from `@woocommerce/helix` any time, that should get it in our reviewer queue.

ismaeldcom committed:
Rebase conflicts

ismaeldcom committed:
Update tests

ismaeldcom committed:
Minor refactor and tests update

Rename `maybe_add_*` methods to `get_*` as they weren't actually adding anything.

ismaeldcom committed:
Fix tests fro WP 6.1

ismaeldcom committed:
Add changelog entry

ismaeldcom committed:
Update notification note to meet new Stripe requriements

We need to use `processing` prop rather than `next_action` updating the intention class accordingly.

ismaeldcom committed:
Update `process_webhook_payment_intent_failed` to support intent with no charges.

Using `last_payment_error` as the source of details. And ensure we return the Stripe error message when there is no error code match.

ismaeldcom committed:
Fix psalm errors

ismaeldcom committed:
Minor comment update

ismaeldcom commented:
Hey @Automattic/helix! We finally have the PR ready for review. I've made some updates to the code as a lot of things have changed since the initial implementation.

I would like to know if we need to consider anything else for WCPay Subscriptions and if there is a way to test manual renewals like in WC Subs.

Thanks!

oaratovskyi commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -791,4 +794,112 @@ private function is_wcpay_subscription_renewal_order( WC_Order $renewal_order )
+	public function get_mandate_params_for_order( WC_Order $order ): array {
+		$result = [];
+
+		if ( $this->is_subscriptions_enabled() ) {
```
nitpick: we can return early here `if ( ! $this->is_subscriptions_enabled() )`, but this implementation is also good!

ismaeldcom replied:
Good catch! I missed it, I always prefer early returns too 🙂  Addressed in d220c6d

oaratovskyi commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -621,45 +629,41 @@ private function get_order_from_event_body_intent_id( $event_body ) {
-				$failure_message = __( 'You have exceeded the number of allowed verification attempts.', 'woocommerce-payments' );
-				break;
-
-			default:
```
I wonder if we need here the `default:` case with `The payment was not able to be processed` failure message? I think it might be helpful if Stripe adds another code that we are not aware of

ismaeldcom replied:
It's actually covered with [this new return](https://github.com/Automattic/woocommerce-payments/blob/d26264a205f503b5691c07e263302117c7f35380/includes/class-wc-payments-webhook-processing-service.php#L667). And instead of returning just `The payment was not able to be processed` I also added Stripe's error message.

Maybe it's a bit confusing because I rewrote the switch to use `return` rather than `break`. So using a single return at the end will act as the `default`. I prefer it when we only want to return something, and especially if we have a nested `switch`.

oaratovskyi replied:
Aaaah, agree, makes sense.

oaratovskyi commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -263,6 +263,7 @@ public function create_and_confirm_intention(
 	 * @param string      $capture_method  - optional capture method (either `automatic` or `manual`).
 	 * @param array       $metadata        - A list of intent metadata.
 	 * @param string|null $customer_id     - Customer id for intent.
+	 * @param array       $additional_parameters  - An array of any additional request parameters, particularly for additional payment methods.
```
nitpick: I see that doc comments are indented in three columns above, this breaks the rule 😅  But again, it's not important, the linter is not complaining

ismaeldcom replied:
Thanks for noticing it, I would like it to be fixed by PHPCBF! Addressed in d220c6d

oaratovskyi reviewed (approved):
@ismaeldcom first of all - big thanks for working on this one! I know it wasn't an easy one and it was blocked a couple of times, great to see it's ready now 🎉 

I followed testing instructions very accurately (classic + UPE). I can confirm that it works as expected: successful and failed renewals for different reasons. I also noticed that after successful renewal another one is scheduled depending on the subscription frequency. 

The code is clean and straightforward, documented very well, and covered by tests, I only described a few nitpicks, overall the code is looking great 💪 

I know that it's still open whether it needs additional work regarding WCPay Subscriptions, but I am pre-approving the PR and looking forward to feedback from the Helix team.

ismaeldcom committed:
Use early return and align docblock params

haszari commented:
FYI @shendy-a8c  - this needs a Helix review, maybe already on your radar?

shendy-a8c committed:
Merge branch 'develop' into fix/2935-recurring-transactions-indian-ccs

shendy-a8c commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -807,4 +810,113 @@ private function is_wcpay_subscription_renewal_order( WC_Order $renewal_order )
+	 * @param WC_Order $renewal_order The subscription renewal order.
+	 * @return array Param to be included or empty array.
+	 */
+	public function get_mandate_param_for_renewal_order( WC_Order $renewal_order ): array {
```
In `get_mandate_params_for_order()` you check if `is_subscriptions_enabled()`, I think it's a good idea to also check here in `get_mandate_param_for_renewal_order()` as well.

Very likely `is_subscriptions_enabled()` returns true because `woocommerce/subscriptions-core` is [packaged](https://github.com/Automattic/woocommerce-payments/blob/c6154cddf73f49e486ee9316da1075c9cf982418/composer.json#L32) in WCPay but just to be safe because otherwise calling `wcs_get_subscriptions_for_renewal_order()` will return a fatal error.

ismaeldcom replied:
I wonder if it's necessary when we will only call `get_mandate_param_for_renewal_order()` within `scheduled_subscription_payment` 🤔 I assume it could only be called if Subscriptions are enabled.

In the other hand `get_mandate_params_for_order()` could be called for an order that's not a subscription.

WDYT?

shendy-a8c replied:
That makes sense! No need to check then.

shendy-a8c commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -276,7 +277,8 @@ public function create_intention(
 		$capture_method = 'automatic',
 		array $metadata = [],
-		$customer_id = null
+		$customer_id = null,
```
I don't see any `create_intention()` calls that is passed by `$additional_parameters`. Do I miss anything?

ismaeldcom replied:
Good catch! It was used in a previous implementation and I forgot to revert it. Addressed in 2777b24

shendy-a8c commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -371,7 +374,7 @@ private function process_webhook_payment_intent_failed( $event_body ) {
 		// Get the order and make sure it is an order and the payment methods match.
 		$order             = $this->get_order_from_event_body_intent_id( $event_body );
-		$payment_method_id = $charges_data['payment_method'] ?? null;
+		$payment_method_id = $payment_method['id'] ?? null;
```
Same with my above comment. I don't understand this change.

I see that originally the value of `$payment_method_id` is from `data.object.charges.data[0].payment_method` and now it's from `data.object.last_payment_error.id`.

shendy-a8c commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -357,10 +357,13 @@ private function process_webhook_payment_intent_amount_capturable_updated( $even
+		$charge_id           = $event_body['data']['object']['charges']['data'][0]['id'] ?? '';
+		$last_payment_error  = $event_body['data']['object']['last_payment_error'] ?? null;
+		$payment_method      = $last_payment_error['payment_method'] ?? null;
+		$payment_method_type = $payment_method['type'] ?? null;
```
I see that originally the value of `$payment_method_type` is from `data.object.charges.data[0].payment_method_details.type` and now it's from `data.object.last_payment_error.type`.

I don't understand why that change. Will we always get `data.object.last_payment_error` for `payment_intent.payment_failed` webhook event despite the failing payment method (`card`, `us_bank_account`, or `au_becs_debit`)?

Will this change work when payment intent failed for `us_bank_account` and `au_becs_debit`? I don't know how to test that. Do you know?

ismaeldcom replied:
Reply to this and:
> I see that originally the value of $payment_method_id is from data.object.charges.data[0].payment_method and now it's from data.object.last_payment_error.id.

I mentioned it on the PR description, but I'll give more context here 🙂 

There are some `payment_intent.payment_failed` that doesn't have any charges. With our current implementation that will result in an on hold order with no additional notes. Instead, we can rely on `data.object.last_payment_error` which contains the error reason, among other things.

There is no existing issue, but we've seen similar cases – a recent Slack conversation about it p1666137895802399/1665973745.206889-slack-CGGCLBN58

It will also work for `us_bank_account` and `au_becs_debit`. As an example [here is a BECS failed intent](https://dashboard.stripe.com/test/connect/accounts/acct_1M4juUQoftj2Jyyc/payments/pi_3M4jyVQoftj2Jyyc17kpQBwj).

To test BECS you'll need to onboard an AU Stripe account, and manually request BECS capability through Stripe's dashboard before being able to enable it in WCPay Settings.

My original idea was to let this part out of the PR and move it to a follow-up. But I think having failed orders on-hold is not an acceptable tradeoff.

shendy-a8c replied:
Thank you for explaining all that. Good to know that `data.object.last_payment_error` will always be available and it's a better alternative than reading from charges.

shendy-a8c commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -807,4 +810,113 @@ private function is_wcpay_subscription_renewal_order( WC_Order $renewal_order )
+	 * @param WC_Order $renewal_order The subscription renewal order.
+	 * @return array Param to be included or empty array.
+	 */
+	public function get_mandate_param_for_renewal_order( WC_Order $renewal_order ): array {
```
Is this intended that the function name is `get_mandate_param_for_renewal_order` instead of `get_mandate_params_for_renewal_order`?

Just a `nit` because it is not symmetrical with `get_mandate_params_for_order`.

ismaeldcom replied:
Yep, I've used the word in singular here because we're only adding a single parameter. Do you think it will sound better if we use the plural in both of them?

shendy-a8c replied:
As long as intended and not an oversight, it's good.

shendy-a8c reviewed (commented):
Is it intended to always pass mandate params (from `get_mandate_params_for_order()` and `get_mandate_param_for_renewal_order()`) even though payment is using a card not issued by India bank?

I was expecting a check if card is issued by India bank and only then pass the mandate parameters but maybe checking which bank country issues the card isn't possible so we have to always pass mandate params to Stripe and Stripe will return mandate id when we receive the `payment_intent.succeeded` webhook event.

shendy-a8c commented:
Is mandate tied to specific card? If someone knows a mandate id, can they use it to create a payment intent with different card and by pass the needed customer confirmation?

shendy-a8c commented:
> I would like to know if we need to consider anything else for WCPay Subscriptions and if there is a way to test manual renewals like in WC Subs.

Instruction to test WCPay Subscription renewal can be found here: PCYsg-DQ4-p2.

shendy-a8c commented:
I saw this PR branch was pretty behind `develop`, so I merged develop into `fix/2935-recurring-transactions-indian-ccs`. I hope it's OK.

shendy-a8c commented:
> Known issues
Changing payment method on a subscription which have a mandate will result in a failed renovation. Using another Indian CC will result in an error by mandate not matching the payment method. And using any other card will also fail because we will attach the initial mandate.
Possible fix: Update the mandate on payment method change, removing it if it's empty.

Will this be addressed, maybe in different PR? If so, please create a separate issue so it is not forgotten.

Changing payment method is part of the critical flows ([ref](https://github.com/Automattic/woocommerce-payments/wiki/WC-Pay-Subscriptions-critical-flows-testing-instructions#change-payment-method-card-for-an-active-subscription)) but I bet it won't be detected as bug because GlobalStep likely will test using US card.

ismaeldcom committed:
Revert unneeded change

There is no need for $additional_parameters in create_intention

ismaeldcom commented:
Thanks for the detailed review @shendy-a8c! 

>Is it intended to always pass mandate params (from get_mandate_params_for_order() and get_mandate_param_for_renewal_order()) even though payment is using a card not issued by India bank?
>
>I was expecting a check if card is issued by India bank and only then pass the mandate parameters but maybe checking which bank country issues the card isn't possible so we have to always pass mandate params to Stripe and Stripe will return mandate id when we receive the payment_intent.succeeded webhook event.

As far as I know, the is no way to check if the card is issued by an Indian bank. Although, now that you mention it, we can consider passing them only for orders using INR currency. Stripe will only return a mandate ID for those cases that require it.

>Is mandate tied to specific card? If someone knows a mandate id, can they use it to create a payment intent with different card and by pass the needed customer confirmation?

The mandate is tied to both the card and subscription. Using an existing mandate ID in an intent with a different card it will fail and show the `Not matching mandate` error shown on the PR description.

>Instruction to test WCPay Subscription renewal can be found here: PCYsg-DQ4-p2.

Thanks! I'll test it and update the PR accordingly soon.

>I saw this PR branch was pretty behind develop, so I merged develop into fix/2935-recurring-transactions-indian-ccs. I hope it's OK.

Sure! 🙂 

> Will this be addressed, maybe in different PR? If so, please create a separate issue so it is not forgotten.
>
>Changing payment method is part of the critical flows ([ref](https://github.com/Automattic/woocommerce-payments/wiki/WC-Pay-Subscriptions-critical-flows-testing-instructions#change-payment-method-card-for-an-active-subscription)) but I bet it won't be detected as bug because GlobalStep likely will test using US card.

Yes, I plan to create a followup issue for both of the known issues before merging this. The PR has been blocked for more than a year, and delaying it again to fix those issues could end up in another big refactor due to planned changes in the affected code. I think it's safe to proceed with it, as it will at least allow renewing subscriptions without errors.

shendy-a8c commented:
> As far as I know, the is no way to check if the card is issued by an Indian bank. Although, now that you mention it, we can consider passing them only for orders using INR currency. Stripe will only return a mandate ID for those cases that require it.

Cool. I was just checking. I don't think there is any change needed here.

> Although, now that you mention it, we can consider passing them only for orders using INR currency.

I think checking for currency is not ideal because the requirement is on the card itself, ie if it's issued by India bank, despite the currency, right?

shendy-a8c committed:
Merge branch 'develop' into fix/2935-recurring-transactions-indian-ccs

ismaeldcom commented:
>I think checking for currency is not ideal because the requirement is on the card itself, ie if it's issued by India bank, despite the currency, right?

Yep, Stripe mentions `Currently, all issuing banks only support INR for subscriptions.` but that could change in the future. So better to keep it as is.

ismaeldcom commented:
Hey @shendy-a8c! Now that every comment is resolved, ~~I'm going to create the follow-up issues~~ done, and proceed with the merge. Anything else on your side before proceeding?

#### Follow-up issues:
- [Improve recurring payments order notes on Indian CCs](https://github.com/automattic/woocommerce-payments/issues/3075#top)
- [Handle changing payment method for orders with a mandate](https://github.com/Automattic/woocommerce-payments/issues/5249#top)

After testing early renewals using WCPay Subscriptions it seems to work as expected. But as it relies on Stripe Billing the mandate is not attached to the intent. So I'm thinking to create another issue to confirm, by asking Stripe first, if we also need to do something in our end for [India recurring payments](https://stripe.com/docs/india-recurring-payments) using the Billing API.

If you agree with the mentioned idea, would you mind creating the WCPay Subscription issue? I'm sure you have way more knowledge about how it works than me, and could add a better description 🙂

shendy-a8c committed:
Merge branch 'develop' into fix/2935-recurring-transactions-indian-ccs

shendy-a8c commented:
Sounds like a plan, @ismaeldcom.

> After testing early renewals using WCPay Subscriptions it seems to work as expected.

I think 'early renewal' is not the term you meant here. WCPay Subscriptions does not support early renewals ([ref](https://woocommerce.com/document/payments/subscriptions-basics/comparison/#feature-comparison)). With WooCommerce Subscriptions, early renewal is possible. Early renewal means one (customer or merchant) can renew the subscriptions before the expected renewal date. Testing WCPay Subscriptions with billing clock even though feels like renewing the subscriptions before the expected renewal date, it is not an early renewal because it is just a way to move the time forward artificially.

> But as it relies on Stripe Billing the mandate is not attached to the intent.

Any link to the payment intent you tested? It might be because you tested with billing clock.

> would you mind creating the WCPay Subscription issue?

I like to make sure it's really a problem by testing without billing clock. More details can be found here: pdjTHR-21b-p2.

As of this PR, I think it looks good and I think it's a good idea to wrap up and follow up with new issues you created and one that I might create after I can confirm there is a problem with WCPay Subscriptions.

ismaeldcom commented:
>I think 'early renewal' is not the term you meant here

You're right! I didn't choose the best name here 🙂  I meant what you explained, using the test billing clock.

>Any link to the payment intent you tested? It might be because you tested with billing clock.

Sure, here is one example: https://dashboard.stripe.com/test/connect/accounts/acct_1Lwn9BQrnu7Ppo4o/payments/pi_3MByt5Qrnu7Ppo4o1TeYLfEs

And yes, I think that's the cause. If I understood the way it works, it creates a test subscription (with a test customer) under the hood. So I think it's expected to not attach the mandate.

Thanks approving the PR and organizing the P2 with further testing for WCPay Subscriptions 👍

shendy-a8c commented:
I just created https://github.com/Automattic/woocommerce-payments/issues/5311 to fix WCPay Subscriptions renewals with Indian CC.
